rfcs/0156-no-direct-pushes.md

```diff
@@ -0,0 +1,173 @@
+---
+feature: no-direct-pushes
+start-date: 2023-07-21
+author: Silvan Mosberger
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Require pull requests for all Nixpkgs commits.
+
+# Motivation
+[motivation]: #motivation
+
+There are currently 197 people with commit access to Nixpkgs, all of whom can push directly to any branch without a pull request.
+Such pushes generally do not notify anybody and do not trigger CI.
+
+This makes such commits susceptible to:
+- Be anonymous
```

Comment:
Isn't «X authored and Y committed» exactly the GitHub way to record who did a direct push??

Reply:
The commit author and committer is decided by the person creating the commit, it can be arbitrarily faked, here's an example of a commit I pushed directly to a test repo with an anonymous author and committer: https://github.com/infinisil/github-test/commit/0553a1afe8ee38d45ef38c7055a7b6c3ee08f3d3

Reply:
Oh wow. There goes my idea of GitHub doing something sensible for once. Thanks for the demonstration.

Reply:
That has more to do with the fact that those are git concepts, so it's not really something in GitHub's control.

Without signed commits, it is also possible for people to impersonate each other too.

Reply:
Given that GitHub's backend is known not to be git, and known to do various non-trivial stuff, storing ans showing who has pushed when this person does not seem to be neither the author nor the committer is perfectly within their control.

Reply:
I initially thought this wouldn't be possible, but actually it should be: When you push, you connect to GitHub's SSH server with a public key that's tied to an authorized GitHub account. This data needs to be stored and presented for each commit though, which would be best if built into GitHub itself.

In fact, the [push webhook event](https://docs.github.com/en/webhooks-and-events/webhooks/webhook-events-and-payloads#push) includes a `sender` field that seemingly contains exactly that, the user that pushed it, which should also allow a third-party to implement such a feature!

Reply:
I now added these points to the RFC in https://github.com/NixOS/rfcs/pull/156/commits/becc35e5b64712c15df0fc65c3402f17bf67f351